Explainer on user-defined operators (#202)
ethereum/solidity-blog: _posts/2023-02-01-user-defined-operators.md

---
layout: post
published: true
title: 'Feature Deep-Dive: User-Defined Operators'
date: '2023-02-01'
author: Kamil Śliwak
category: Explainers
---

> The prominent new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)
> is the support for defining operators on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
> Let's take a closer look into the details of the the new mechanism.

Reviewer: ```suggestion
The prominent new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)
is the support for defining operators on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).
A UDVT is a type of abstraction that allows creating a zero-cost representation of an elementary value type. 
This is similar to creating an alias, but with stricter type requirements.

Bellow, we're going to take a closer look into the details of this new mechanism.
```

### Built-in operators
Solidity provides the following operators for its built-in types:

| Category            | Operators                                                               |
|---------------------|-------------------------------------------------------------------------|
| Bitwise             | `&`, `\|`, `^`, `<<`, `>>`, `~`                                         |
| Arithmetic          | `+`, `-`, `*`, `/`, `%`, `**`                                           |
| Comparison          | `==`, `!=`, `<`, `<=`, `>`, `>=`                                        |
| Boolean             | `!`, `&&`, `||`                                                         |
| Assignment          | `=`, `+=`, `-=`, `>=`, `*=`, `/=`, `%=`, `&=`, `|=`, `^=`, `<<=`, `>>=` |
| Increment/decrement | `++`, `--`                                                              |
| Ternary             | `?:`                                                                    |
| Other               | `delete`, `new`                                                         |

> Most of these operators work only with [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
> 
> In particular, with the exception of `?:` and `=`, these operators are not available on user-defined
> value types, making their use much less convenient than the built-in ones.
> One way to make the use of these types more convenient is attaching helper functions with
> [`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).

Reviewer: ```suggestion
Most of these operators work only with [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).
With the exception of `?:` and `=`, all of these operators are not available on user-defined
value types, making their use much less convenient than the built-in ones.
One way to make the use of these types more convenient is attaching helper functions with
[`using for`](https://docs.soliditylang.org/en/latest/contracts.html#using-for).
```

> ### Quick Refresher on `using for`
> `using for` can attach a function to a type.
> 
> ```solidity
> type Int is int;
> using {add} for Int;
> 
> function add(Int a, Int b) pure returns (Int) {
>     return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
> }
> 
> function test(Int a, Int b) pure returns (Int) {
>     return a.add(b);
> }
> ```
> 
> Currently it has 3 forms:
> 1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
> 1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
>     cannot be named in the language (e.g. types of literals or array slices).
> 1. `using {f, g, h} for T`: Attaches functions `f`, `g` and `h` to type `T`.
> 
> `using for` can be used at file level or inside contracts, including libraries (but not interfaces).
> 
> At file level the directive can be marked as `global` which extends its effect to all source units
> where a type is visible.
> The use of `global` is possible only with UDVTs, structs and enums that are defined in the current source unit.
> 
> The following example shows off the main features of `using for` in combination wth UDVTs:
> 

Reviewer: ```suggestion
### Quick Refresher on `using for`
The directive `using A for B;` can be used to attach functions (A) as member functions to any type (B). 
These functions will receive the object they are called on as their first parameter.

```solidity
type Int is int;
using {add} for Int;

function add(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
}

function test(Int a, Int b) pure returns (Int) {
    return a.add(b);
}
```

Currently, it has 3 forms:
1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
    cannot be named in the language (e.g. types of literals or array slices).
1. `using {f, g, h} for T`: Attaches functions `f`, `g`, and `h` to type `T`.

✅ Can be used:
- at the file level
- inside contracts,  including libraries

❌ Can not be used
- Inside contracts with interfaces

At the file level, the directive can be marked as `global`, which extends its effect to all source units
where a type is visible.
The use of `global` is possible only with UDVTs, structs, and enums that are defined in the current source unit.

The following example shows off the main features of `using for` in combination with UDVTs:

```

> ### Introducing User-Defined Operators
> The new feature adds another form of the `using for` syntax, where instead of attaching a function,
> it is assigned an operator symbol.
> The function is then invoked whenever the operator is used on the values of the type:
> 
> ```solidity
> type Int is int;
> using {add as +} for Int;
> 
> function add(Int a, Int b) pure returns (Int) {
>     return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
> }
> 
> function test(Int a, Int b) pure returns (Int) {
>     return a + b;
> }
> ```
> 
> Note that defining an operator for a type does not automatically make it available as an attached function.
> E.g. `using {add as +} for Int` does not imply `using {add} for Int`.
> 
> #### Which Operators Can Be User-Defined?
> | Category   | Operators                        |
> |------------|----------------------------------|
> | Bitwise    | `&`, `\|`, `^`,  `~`             |
> | Arithmetic | `+`, `-`, `*`, `/`, `%`          |
> | Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |
> 
> Of the operators listed above `!` and `~` are unary, `-` can be both unary and binary and the
> rest is always binary.
> Note that Solidity does not have binary `+`.
> 
> Several operators have been excluded from the initial implementation for the sake of simplicity.
> We are considering enabling them in one of the future releases if the design issues can be
> resolved satisfactorily:
> 1. `&&`, `||` and `!` are not necessary for most UDVTs because user-defined comparison operators
>     must return `bool`, for which the built-in boolean operators are sufficient.
> 
>     In case of `&&` and `||`  an additional complication is their special short-circuiting behavior.
> 1. `<<`, `>>` and `**` have non-uniform arguments.
>     For built-in types the right-hand-side is always an unsigned type.
> 1. `=` is one of the few operators that already work with UDVTs.
>     The new feature is not meant to allow redefining existing operators.
> 1. Assignment and increment/decrement operators in general modify their argument and return a reference.
> 
>     `++` and `--` also have separate prefix and postfix variants with different semantics.
> 

Reviewer: ```suggestion
### Introducing User-Defined Operators
The new feature adds another form of the `using for` syntax, where instead of attaching a function,
it is assigned an operator symbol.
The function is then invoked whenever the operator is used on the values of the type:

```solidity
type Int is int;
using {add as +} for Int;

function add(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
}

function test(Int a, Int b) pure returns (Int) {
    return a + b;
}
```

Note that defining an operator for a type does not automatically make it available as an attached function.
E.g. `using {add as +} for Int` does not imply `using {add} for Int`.

#### Which Operators Can Be User-Defined?
| Category   | Operators                        |
|------------|----------------------------------|
| Bitwise    | `&`, `\|`, `^`,  `~`             |
| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |

Of the operators listed above, `!` and `~` are unary, `-` can be both unary and binary and the
rest is always binary.
Note that Solidity does not have binary `+`.

Several operators have been excluded from the initial implementation for the sake of simplicity.
We are considering enabling them in one of the future releases if the design issues can be
resolved satisfactorily:
1. `&&`, `||` and `!` are not necessary for most UDVTs because user-defined comparison operators
    must return `bool`, for which the built-in boolean operators are sufficient.
    In case of `&&` and `||`  an additional complication is their particular short-circuiting behavior.
1. `<<`, `>>` and `**` have non-uniform arguments.
    For built-in types, the right-hand side is always an unsigned type.
1. `=` is one of the few operators that already work with UDVTs.
    The new feature is not meant to allow redefining existing operators.
1. Assignment and increment/decrement operators, in general, modify their argument and return a reference.

    `++` and `--` also have a separate prefix and postfix variants with different semantics.

```

> #### Which Types Can Have User-Defined Operators?
> Currently operators can only be defined for user-defined value types.
> We are also considering enabling them for structs and enums in future releases.
> There are no plans to allow defining operators for built-in value types, especially in cases where
> an operator is already provided by the compiler.
> 
> Note that the type wildcard (`*`) is not allowed in `using for` when defining operators, which
> is in line with how it already worked when used attach individual functions.
> 
> Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
> must always be invoked with exact types.
> 

Reviewer: ```suggestion
#### Which Types Can Have User-Defined Operators?
Currently, operators can only be defined for user-defined value types.
We are also considering enabling them for structs and enums in future releases.
There are no plans, for now, to allow defining operators for built-in value types, especially in cases where
an operator is already provided by the compiler.

Note that the type wildcard (`*`) is not allowed in `using for` when defining operators, which
is in line with how it already worked when used to attach individual functions.

Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
must always be invoked with exact types.

```

The return value must be of the same type as well, except for comparison operators, where it must be `bool`.

A few points of note:
- Private library functions are allowed but visible only within the library.
- Built-in functions like `keccak256`, `revert` or `abi.encode()` cannot be used.
- Library functions are allowed but contract and interface functions are not.
    This means that the definition cannot be virtual.
- Overloaded functions are not purposefully disallowed, but due to the lack of a language mechanism
    for selecting a specific overload, it is currently not possible to use such functions.
- Operator definition may be a public library function.
    External functions will eventually be allowed too
    (see [issue #13765](https://github.com/ethereum/solidity/issues/13765)).
- Being a `pure` function, the operator defintion is allowed to make `pure` external calls.

The rules for `using for global` with operators are the same as for bound functions:
it is only allowed at file level, the type must be defined in the current source unit and the
effects extend to wherever the type is visible, not just the current source unit.

> #### Can Operators Be Redefined?
> It is not possible to redefine any built-in operator.
> All valid cases are the ones where previously there was no built-in operator.
> 
> Only one definition is allowed for each user-defined operators for a given type.
> Note that:
> - Unary and binary `-` count as two separate operators.
> - As long as the definition stays the same, it is not an error to repeat the `using for` directive,
>     just as it is not an error to attach the same function to a type twice.
> 
> `using for` is not inherited by derived contracts and this applies to operators as well.
> For this reason, it is possible for a base contract and a derived contract to use different definitions
> for the same operator.
> The same is not possible at file level.
> If an operator is defined at file level, it cannot have a different definition inside a contract where
> that file-level definition is visible.
> 

Reviewer: ```suggestion
#### Can Operators Be Redefined?
It is not possible to redefine any built-in operator.
All valid cases are the ones where previously there was no built-in operator.

As such, only one definition is allowed for each user-defined operator for a given type.
Note that:
- Unary and binary `-` count as two separate operators.
- As long as the definition stays the same, it is not an error to repeat the `using for` directive,
    just as it is not an error to attach the same function to a type twice.

`using for` is not inherited by derived contracts, and this applies to operators as well.
For this reason, it is possible for a base contract and a derived contract to use different definitions
for the same operator.
The same is not possible at the file level.
If an operator is defined at the file level, it cannot have a different definition inside a contract where
that file-level definition is visible.

```

> ### AST Changes
> The new feature introduces two new attributes to the AST:
> 1. Functions listed in `functionList` on `UsingForDirective` nodes, may now optionally have an
>     `operator` attribute.
>     When present, it indicates that the function defines an operator and specifies its name.
> 1. `BinaryOperation` and `UnaryOperation` may now have a `function` attribute.
>     If present, the attribute indicates that the operation is actually an invocation of a user-defined
>     operator and specifies the ID of the function used to define it.
> 
> This effectively introduces a new way to perform a function call, not represented by a `FunctionCall` node.
> This will affect any tool that needs to detect and follow function calls, for example to build a
> control flow graph.

Reviewer: ```suggestion
### AST Changes
The new feature introduces two new attributes to the Abstract Syntax Tree (AST):
1. Functions listed in `functionList` on `UsingForDirective` nodes, may now optionally have an
    `operator` attribute.
    When present, it indicates that the function defines an operator and specifies its name.
1. `BinaryOperation` and `UnaryOperation` may now have a `function` attribute.
    If present, the attribute indicates that the operation is actually an invocation of a user-defined
    operator and specifies the ID of the function used to define it.

This effectively introduces a new way to perform a function call, not represented by a `FunctionCall` node.
This will affect any tool that needs to detect and follow function calls, for example, to build a
control flow graph.
```

---
layout: post
published: true
title: 'Feature Deep-Dive: User-Defined Operators'
date: '2023-02-01'
author: Kamil Śliwak
category: Explainers
---

> The prominent new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)

Reviewer 2: ```suggestion
The prominent new feature in Solidity 0.8.19
```

> is the support for defining operators on [user defined value types](https://docs.soliditylang.org/en/latest/types.html#user-defined-value-types) (UDVTs).

Author: Added an emphasis. I think bolding it would be a bit too strong.

### Built-in operators
Solidity provides the following operators for its built-in types:

| Category            | Operators                                                                |
|---------------------|--------------------------------------------------------------------------|
| Bitwise             | `&`, `\|`, `^`, `<<`, `>>`, `~`                                          |
| Arithmetic          | `+`, `-`, `*`, `/`, `%`, `**`                                            |
| Comparison          | `==`, `!=`, `<`, `<=`, `>`, `>=`                                         |
| Boolean             | `!`, `&&`, `\|\|`                                                        |
| Assignment          | `=`, `+=`, `-=`, `^=`, `*=`, `/=`, `%=`, `&=`, `\|=`, `^=`, `<<=`, `>>=` |
| Increment/decrement | `++`, `--`                                                               |
| Ternary             | `?:`                                                                     |
| Other               | `delete`, `new`                                                          |

Most of these operators work only with [value types](https://docs.soliditylang.org/en/latest/types.html#value-types).

> With the exception of `?:` and `=`, all of these operators are not available on user-defined
> value types, making their use much less convenient than the built-in ones.

Reviewer 2: ```suggestion
With the exception of `?:` and `=`, all of these operators are not available on UDTVs, 
making their use much less convenient than the built-in ones.
```

Author: I'm using them interchangeably for some variety so it was intentional.

These functions can then be used as members on any object of that type and receive the object as
their first parameter.

```solidity
type Int is int;
using {add} for Int;

function add(Int a, Int b) pure returns (Int) {
    return Int.wrap(Int.unwrap(a) + Int.unwrap(b));
}

function test(Int a, Int b) pure returns (Int) {
    return a.add(b);
}
```

Currently, `using for` has 3 forms:
1. `using L for T`: attaches all non-private functions from library `L` to type `T`.
1. `using L for *`: attaches all non-private functions from library `L` to any type, including types that
    cannot be named in the language (e.g. types of literals or array slices).
1. `using {f, g, h} for T`: Attaches functions `f`, `g`, and `h` to type `T`.

✅ Can be used:
- At the file level.
- Inside contract and library definitions.

❌ Can not be used:
- Inside interface definitions.

At the file level, the directive can be marked as `global`, which extends its effect to all source
units where the type is visible.
The use of `global` is possible only with UDVTs, structs, and enums that are defined in the current source unit.

### Introducing User-Defined Operators

> The new feature adds another form of the `using for` syntax, where instead of attaching a function,
> it is assigned an operator symbol.

Reviewer 2: ```suggestion
The new feature in [Solidity 0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18)
adds another form of the `using for` syntax, where a function can be assigned to an operator symbol.
```

#### But Why?
The primary use case we had in mind when extending UDVTs was support for fixed-point numbers.

Fixed-point types were originally going to be an integral part of the language, and the
partially implemented [`ufixed` and `fixed`](https://docs.soliditylang.org/en/latest/types.html#fixed-point-numbers)
types are the vestiges of that old design.
While trying to agree the final design for the feature, however, we realized that no implementation
is going to be a trade-off good enough for most situations and cannot become a suitable default choice.
We decided to leave the choice up to library authors and instead focus on providing primitives
necessary to build their own fixed-point types that feel like a native type.
User-defined operators are a step towards that goal.

#### Which Operators Can Be User-Defined?
| Category   | Operators                        |
|------------|----------------------------------|
| Bitwise    | `&`, `\|`, `^`,  `~`             |
| Arithmetic | `+`, `-`, `*`, `/`, `%`          |
| Comparison | `==`, `!=`, `<`, `<=`, `>`, `>=` |

Of the operators listed above, `~` is unary, `-` can be both unary and binary and the rest is always binary.

> Note that Solidity does not have binary `+`.

Reviewer 2: ```suggestion
Note that Solidity does not have unary `+`.
```

Author: Nice catch!

#### Which Types Can Have User-Defined Operators?
Currently, operators can only be defined for user-defined value types.
We are also considering enabling them for structs and enums in future releases.
There are no plans, for now, to allow defining operators for built-in value types, especially in cases where
an operator is already provided by the compiler.

Note that the type wildcard (`*`) is not allowed in `using for` when defining operators, which
is in line with how it already worked when used to attach individual functions.

Since currently there are no implicit conversions allowed to and from UDVTs, user-defined operators
must always be invoked with exact types.

#### What Are the Restrictions on User-Defined Operator Definitions?

> The operator definition must be a pure function, defined at the file level or a library.

Reviewer 2: ```suggestion
The operator definition must be a pure function, defined at the file level or inside a library.
```

> All parameters must be of the same type, which must be the type specified in `using for`.

Reviewer 2: ```suggestion
All parameters must be of the same type specified in `using for`.
```

Author: I'll try to reword it, but the proposed version sounds a bit ambiguous to me. As if the parameters were defined in `using for` :)

The return value must be of the same type as well, except for comparison operators, where it must be `bool`.

A few points of note:
- Being a `pure` function, the operator definition is allowed to make `pure` external calls.
- Library functions are allowed but contract and interface functions are not.
    This means that the definition cannot be virtual.
- Private library functions are allowed but visible only within the library.
- Operator definition may be a public library function.
    External functions will eventually be allowed too
    (see [issue #13765](https://github.com/ethereum/solidity/issues/13765)).
- Overloaded functions are not purposefully disallowed, but due to the lack of a language mechanism
    for selecting a specific overload, it is currently not possible to use such functions.
- Built-in functions like `keccak256`, `revert` or `abi.encode()` cannot be used even if their

>     parameter and return value types happen to match the type.

Reviewer 2: ```suggestion
    parameter and return value types happen to match the type specified in `using for`.
```

> The rules for `using for global` with operators are the same as for bound functions:

Reviewer 2: ```suggestion
The rules for `using for global` with operators are the same as for attached functions:
```

literals that are already available in the language:

```solidity
type UFixed16x2 is uint16;

function uf(int16 mantissa, uint8 exponent) pure suffix returns (UFixed16x2) { ... }
```
```solidity
UFixed16x2 circumference = 2 uf * 3.14 uf * radius;
```

While the application of a suffix will involve a function call, eventually, with the help of
compile-time evaluation, literal suffixes are going to become as efficient as native literals.

#### More Operators
Several operators have been excluded from the initial implementation for the sake of simplicity:
1. `&&`, `||` and `!` are not necessary for most UDVTs because user-defined comparison operators
    must return `bool`, for which the built-in boolean operators are sufficient.
    In case of `&&` and `||`  an additional complication is their short-circuiting behavior.
1. `<<`, `>>` and `**` have non-uniform arguments.
    For built-in types, the right-hand side is always an unsigned type.
1. `=` is one of the few operators that already worked with UDVTs.
    The new feature is not meant to allow redefining existing operators so providing a custom
    definition for it will not be possible.

> 1. Other assignment operators modify their argument and return a reference.

Reviewer 2: ```suggestion
1. Other assignment operators, as `+=`, for example, modify their argument and return a reference.
```

    There are currently no references to value types in the language which makes implementing them
    directly impossible.
    It may also be desirable to have the compiler provide them automatically whenever the
    corresponding bitwise or arithmetic operator is defined on the type.
1. `++` and `--` modify their argument and also have a separate prefix and postfix variants with
    different semantics.

> We are considering enabling them in one of the future releases, provided that the design issues can
> be resolved satisfactorily.

Reviewer 2: ```suggestion
We are considering enabling them, (except `&&`, `||` and `!`, as mentioned before )
in one of the future releases, provided that the design issues can be resolved satisfactorily.
```

Author: I mean, that includes also also `=`. I wanted the list above to indicate which those are instead of repeating them. I'll see if I can reword it.

There are use cases where operators like `++`, `--`, or `**` would be a natural fit, so they are
very likely to become available sooner or later.

#### Operators on Structs
While UDVTs were the primary target, not all types can reasonably fit into a single slot.
The way to solve this problem is to allow operators taking and returning structs.
The main thing holding us back are the planned changes to reference types, in particular the fact
that data location (`storage`, `memory`, etc.) is going to become a part of the type itself.
To avoid having to introduce breaking changes later, we decided to defer support for structs until

> these changes are behind us.

Reviewer: ```suggestion
these changes are behind us.

#### Why should you use this
@matheusaaguiar ?
```

Reviewer 2: Working on this.
